### PR TITLE
test-network: fix test_bond_operstate

### LINKED ISSUE #12344: `test_bond_operstate` from systemd-networkd-tests.py is flaky
In past several runs I noticed fails in `test_bond_operstate`:

```
======================================================================
FAIL: test_bond_operstate (__main__.NetworkdNetWorkBondTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 1573, in test_bond_operstate
    self.assertRegex(output, 'State: no-carrier \(configured\)')
AssertionError: Regex didn't match: 'State: no-carrier \\(configured\\)' not found in '● 104: bond99\n       Link File: /usr/lib/systemd/network/99-default.link\n    Network File: /run/systemd/network/bond99.network\n            Type: bond\n           State: degraded-carrier (configured)\n          Driver: bonding\n      HW Address: 3e:34:da:ee:33:a4\n         Address: 192.168.123.45\n                  fe80::3c34:daff:feee:33a4'
```

Full log: https://ci.centos.org/job/systemd-pr-build/lastFailedBuild/artifact/systemd-centos-ci/artifacts_t2Zcvj/testsuite-logs-upstream.BXq/systemd-networkd-tests.py_FAIL.log

Affected jobs:
https://ci.centos.org/job/systemd-pr-build/5796/ (PR #12341)
https://ci.centos.org/job/systemd-pr-build/5783/ (PR #12334)
https://ci.centos.org/job/systemd-pr-build/5770/ (PR #12326)

### PR DESCRIPTION
**Pull request by yuwata**
Hopefully fixes #12344.

### CODE REVIEW

**Comment by ssahani** [comment]
cc @mrc0mmand  @evverx .

**Comment by mrc0mmand** [comment]
@yuwata thank you!

Looks promising, but this needs to be fixed:

```
======================================================================
ERROR: test_address_preferred_lifetime_zero_ipv6 (__main__.NetworkdNetWorkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 1090, in test_address_preferred_lifetime_zero_ipv6
    self.check_operstate('dummy98', 'routable')
  File "./test/test-network/systemd-networkd-tests.py", line 204, in check_operstate
    self.assertRegex(self.get_operstate(link, show_status, setup_state), expected)
  File "/usr/lib64/python3.6/unittest/case.py", line 1311, in assertRegex
    if not expected_regex.search(text):
TypeError: expected string or bytes-like object

----------------------------------------------------------------------
Ran 79 tests in 1034.879s

FAILED (errors=1, expected failures=8)
```

**Comment by yuwata** [comment]
Updated. Fixed only the above error.

**Comment by yuwata** [comment]
Now CentOS CI passes.
@mrc0mmand Could you check the flakiness of the test?

**Comment by mrc0mmand** [comment]
@yuwata thanks! Will give it few (dozen) tries in systemd/systemd-centos-ci#109.

**Comment by mrc0mmand** [comment]
After seven runs I got this:

```
======================================================================
FAIL: test_bond_operstate (__main__.NetworkdNetWorkBondTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 1584, in test_bond_operstate
    self.assertTrue(bond_has_no_carrier)
AssertionError: False is not true
```

https://ci.centos.org/job/systemd-ci-build/471/artifact/artifacts_5NlKGN/testsuite-logs-upstream.QgE/systemd-networkd-tests.py_6.log

**Comment by yuwata** [comment]
@mrc0mmand Could you restart the stress test?

@ssahani Could you review the last commit?

**Comment by mrc0mmand** [comment]
13 iterations without any issue so far, looks promising! https://ci.centos.org/job/systemd-ci-build/472/console

File: test/test-network/systemd-networkd-tests.py [hunk 1]
```diff
@@ -1554,63 +1546,42 @@ def test_bond_operstate(self):
...
-        print(output)
-        self.assertRegex(output, 'State: routable \(configured\)')
+        self.check_operstate('dummy98', 'enslaved')
+        self.check_operstate('test1', 'enslaved')
+        self.check_operstate('bond99', 'routable')
 
         self.assertEqual(subprocess.call(['ip', 'link', 'set', 'dummy98', 'down']), 0)
         self.assertEqual(subprocess.call(['ip', 'link', 'set', 'test1', 'down']), 0)
-        time.sleep(5)
+        time.sleep(2)
```
Comment (ssahani) [inline test/test-network/systemd-networkd-tests.py:1569]: why change the sleep.?

File: test/test-network/systemd-networkd-tests.py [hunk 2]
```diff
@@ -1554,63 +1546,42 @@ def test_bond_operstate(self):
...
-        print(output)
-        self.assertRegex(output, 'State: off \(configured\)')
+        self.check_operstate('dummy98', 'off')
+        self.check_operstate('test1', 'off')
 
-        output = subprocess.check_output(['networkctl', 'status', 'test1']).rstrip().decode('utf-8')
-        print(output)
-        self.assertRegex(output, 'State: off \(configured\)')
+        bond_has_no_carrier=False
+        for trial in range(30):
```
Comment (ssahani) [inline test/test-network/systemd-networkd-tests.py:1575]: So we trying to 30 sec at most ?

File: src/network/networkd-link.c [hunk 3]
```diff
@@ -1637,7 +1637,8 @@ bool link_has_carrier(Link *link) {
 
         if (link->kernel_operstate == IF_OPER_UNKNOWN)
                 /* operstate may not be implemented, so fall back to flags */
-                if ((link->flags & IFF_LOWER_UP) && !(link->flags & IFF_DORMANT))
+                if (FLAGS_SET(link->flags, IFF_LOWER_UP | IFF_RUNNING) &&
+                    !FLAGS_SET(link->flags, IFF_DORMANT))
                         return true;
```
Comment (ssahani) [inline src/network/networkd-link.c:1642]: This look Good.

File: test/test-network/systemd-networkd-tests.py [hunk 4]
```diff
@@ -220,6 +220,19 @@ def wait_online(self, links_with_operstate, timeout='20s', bool_any=False):
             args += ['--any']
         subprocess.check_call(args)
 
+    def get_operstate(self, link, show_status=True, setup_state='configured'):
+        output = subprocess.check_output(['networkctl', 'status', link]).rstrip().decode('utf-8')
```
Comment (keszybz) [inline test/test-network/systemd-networkd-tests.py]: <strike>`subprocess.check_output(..., text=True).rstrip()` instead?</strike>

`subprocess.check_output(..., universal_newlines=True).rstrip()` instead (for python3.5 compat)?

(Calling rstrip() before decode() is conceptually wrong, even though there's no difference in this case...)

File: test/test-network/systemd-networkd-tests.py [hunk 5]
```diff
@@ -220,6 +220,19 @@ def wait_online(self, links_with_operstate, timeout='20s', bool_any=False):
...
+        if show_status:
+            print(output)
+        for line in output.split('\n'):
```
Comment (keszybz) [inline test/test-network/systemd-networkd-tests.py]: output.splitlines()

**Comment by mrc0mmand** [comment]
After seven hours and 22 iterations later, the machine got forcefully returned back to the pool, but the test itself passed in all cases. Will try to re-run it, just out of curiosity, but it seems pretty stable :-)

**Comment by evverx** [comment]
> After seven hours and 22 iterations later, the machine got forcefully returned back to the pool, but the test itself passed in all cases. Will try to re-run it, just out of curiosity, but it seems pretty stable :-)

Wow, I seriously doubt any other test in the systemd repository (apart from the fuzzers, which are supposed to be destroyed by throwing gibberish at them :-)) can withstand this kind of stress testing so in terms of stability it looks great to me! @yuwata thank you!

**Comment by mrc0mmand** [comment]
I think I should've left my curiosity unsatisfied, because running the stress test again (twice[0][1]), it failed during the first few iterations with:

```
======================================================================
FAIL: test_bond_operstate (__main__.NetworkdNetWorkBondTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 1584, in test_bond_operstate
    self.assertTrue(bond_has_no_carrier)
AssertionError: False is not true

----------------------------------------------------------------------
Ran 80 tests in 1177.288s
```

Not sure if it helps, but I noticed that the passing run was using *version* `v242-478-g69de308` (as reported by `git describe`) whereas both failing runs were on `v242-479-g0d25abb` (but these object IDs are particularly useless, as they point to merge commits provided by the `merge` ref).

[0] https://ci.centos.org/job/systemd-ci-build/477/artifact/artifacts_4yplHr/testsuite-logs-upstream.iao/systemd-networkd-tests.py_0.log
[1] https://ci.centos.org/job/systemd-ci-build/476/artifact/artifacts__ee4VI/testsuite-logs-upstream.unS/systemd-networkd-tests.py_2.log

File: test/test-network/systemd-networkd-tests.py (hunk 1 above)
Comment (yuwata) [reply test/test-network/systemd-networkd-tests.py:1569]: As we wait for 30 seconds at most in the later loop. It is not necessary to wait a lot here.

File: test/test-network/systemd-networkd-tests.py (hunk 2 above)
Comment (yuwata) [reply test/test-network/systemd-networkd-tests.py:1575]: Yes. It may be too much, but hopefully only few seconds in the most cases.

**Comment by yuwata** [comment]
@mrc0mmand Thank you for testing this so many times. I think the failure seems to be caused by kernel. Not only `networkctl` but `ip` command also shows the bonding interface has carrier, that is, no `NO-CARRIER` in the output. And, on the newer kernel I use `kernel-5.0.16-201.fc29.x86_64`, the test always passes.

So, we need to just adjust the test as it will not fail when the kernel does not recognize the no-carrier state.

**Comment by yuwata** [comment]
@mrc0mmand Updated. Please run the stress test again. Thank you!

**Comment by mrc0mmand** [comment]
@yuwata thanks for the explanation! Started it again, fingers crossed!

File: test/test-network/systemd-networkd-tests.py [hunk 6]
```diff
@@ -1566,11 +1566,27 @@ def test_bond_operstate(self):
...
+                bond_has_no_carrier=True
+                break
+            time.sleep(1)
+
+        if not bond_has_no_carrier:
+            # Huh? Kernel does not recognize that all slave interfaces are down?
+            # Let's confirm that networkd's operstate is consistent with ip's result.
+            output = subprocess.check_output(['ip', 'address', 'show', 'bond99']).rstrip().decode('utf-8')
+            print(output)
+            self.assertNotRegex(output, 'NO-CARRIER')
```
Comment (keszybz) [inline test/test-network/systemd-networkd-tests.py:1585]: Use `else:`, and avoid one sleep:
```python3
for trial in range(30):
     if trial > 0:
            time.sleep(1)
     output = subprocess.check_output(['ip', 'address', 'show', 'bond99'], universal_newlines=True).rstrip()
     print(output)
     if self.get_operstate('bond99') == 'no-carrier':
            break
else:
      # Huh? Kernel does not recognize that all slave interfaces are down?
      # Let's confirm that networkd's operstate is consistent with ip's result.
      self.assertNotRegex(output, 'NO-CARRIER')
```

**Comment by mrc0mmand** [comment]
So, after another seven hours and 22 iterations later, the machine found itself a way back to the pool, as expected, and the test passed in all cases. I decided to not take my chances this time and ran it again. So far it's on its fifth iteration (after an hour and half) and still without any issue. Don't want to jinx it, but looks like this time it might finally work flawlessly :-)

**Comment by yuwata** [comment]
@keszybz Thank you for the comments. Now the all points are addressed.

@mrc0mmand Thank you so much!

**Pull request merged**